### PR TITLE
handle signup route and controller . 

### PR DESCRIPTION
**Pull request by MohammedOmar123**
1- I created a database query to insert users into the database.
2- I created a checkUserExist query to check if the user is already signup. 
3- I created the generate Token function.
4- I created the validateSighup function and I added a custom error message to it.
5- I added test cases for all expected responses.

### CODE REVIEW

**Review by SaraDahman — APPROVED** [review]

File: server/models/users.ts [hunk 1]
```diff
@@ -18,6 +18,7 @@ const User = sequelize.define('User', {
   email: {
     type: DataTypes.STRING(40),
```
Comment (Mu7ammadAbed) [inline server/models/users.ts:19]: I think limiting the email to 40 wouldn't be a good idea

File: server/helpers/ErrorHandler.ts [hunk 2]
```diff
@@ -3,6 +3,8 @@ import { Response, Request, NextFunction } from 'express';
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
 const ErrorHandler = (error, req: Request, res: Response, next: NextFunction) => {
   const { status, message } = error;
+
+  if (error.name === 'ValidationError') return res.status(400).json({ msg: error.message });
```
Comment (Mu7ammadAbed) [inline server/helpers/ErrorHandler.ts:7]:
```suggestion
  if (error.name === 'ValidationError') return res.status(400).json({ message: error.message });
```

File: server/controllers/account/signUp.ts [hunk 3]
```diff
@@ -1,7 +1,28 @@
...
+    const checkResultQuery = await checkUserExist(email);
+
+    if (checkResultQuery) throw new CustomError(400, 'This email is already exist,Please check your email again');
+
+    const hashed = await bcrypt.hash(password, 10);
+    const result = await signupQuery({ ...req.body, password: hashed });
+
+    const token = await generateToken(result.getDataValue('id'), email);
+
+    res.cookie('token', token, { httpOnly: true, secure: true }).json({ msg: 'Your Account Created Successfully' });
```
Comment (Mu7ammadAbed) [inline server/controllers/account/signUp.ts:22]:
```suggestion
    res.cookie('token', token, { httpOnly: true, secure: true }).status(201).json({ message: 'Your Account Created Successfully' });
```

File: server/database/queries/account/signupQuery.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+const checkUserExist = async (email:string) => User.findOne({
+  where: { email },
+});
+const signupQuery = async ({
+  firstName, lastName, email, password,
+}:IUserInfo) => User.create({
+  first_name: firstName,
+  last_name: lastName,
+  email,
+  password,
```
Comment (Mu7ammadAbed) [inline server/database/queries/account/signupQuery.ts:13]:
```suggestion
  hashed_password,
```

File: server/helpers/generateToken.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import jwt from 'jsonwebtoken';
+
+const generateToken = (id:string, email:string) => new Promise((resolve, rejected) => {
```
Comment (Mu7ammadAbed) [inline server/helpers/generateToken.ts:3]:
```suggestion
const generateToken = (id:string, email:string): Promise <string> => new Promise((resolve, rejected) => {
```

File: server/validation/signupValidation.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    }),
+    password: Joi.string().pattern(/^(?!.*[\s])(?=.*[0-9])(?=.*\W)[a-zA-Z0-9\W]{6,15}$/).required().messages({
+      'any.required': 'Password is a required field so be sure to fill it on',
+      'string.empty': 'Password is a required field so be sure to fill it on',
+      'string.pattern.base': 'Password must be at least 6 characters from letters,digits and special characters',
+
+    }),
+    confirmPassword: Joi.any().equal(Joi.ref('password')).required().label('Confirm password')
+      .messages({
+        'any.required': 'confirm Password is a required field so be sure to fill it on',
```
Comment (Mu7ammadAbed) [inline server/validation/signupValidation.ts:30]:
```suggestion
        'any.required': 'Confirm Password is a required field so be sure to fill it on',
```

**Review by SaraDahman — APPROVED** [review]

File: server/controllers/account/signUp.ts [hunk 7]
```diff
@@ -14,12 +14,12 @@ const signup = async (req: Request, res: Response, next: NextFunction) => {
 
     if (checkResultQuery) throw new CustomError(400, 'This email is already exist,Please check your email again');
 
-    const hashed = await bcrypt.hash(password, 10);
-    const result = await signupQuery({ ...req.body, password: hashed });
+    const hashedPassword = await bcrypt.hash(password, 10);
+    const result = await signupQuery({ ...req.body, hashedPassword });
```
Comment (Mu7ammadAbed) [inline server/controllers/account/signUp.ts:18]:
```suggestion
    const result = await signupQuery({ firstName, lastName, email, hashedPassword });
```

File: server/models/users.ts [hunk 8]
```diff
@@ -16,7 +16,7 @@ const User = sequelize.define('User', {
     allowNull: false,
   },
   email: {
-    type: DataTypes.STRING(40),
+    type: DataTypes.STRING(100),
```
Comment (Mu7ammadAbed) [inline server/models/users.ts:19]:
```suggestion
    type: DataTypes.STRING(50),
```

**Review by Mu7ammadAbed — APPROVED** [review]

**Review by hkmusameh01 — APPROVED** [review]

**Review by SaraDahman — APPROVED** [review]

**Pull request merged**